File: day71-80/day79/1593. Split a String Into the Max Number of Unique Substrings.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        maxLength=1
        def backtrack(i, seen):
            nonlocal maxLength
            if i == len(s):
                maxLength = max(maxLength, len(seen))
                return
            for j in range(i+1, len(s)+1):
                if s[i:j] in seen:
                    continue
                seen.add(s[i:j])
                backtrack(j, seen)
                seen.remove(s[i:j])
        
        backtrack(0, seen=set())
        return maxLength
sol=Solution()
print(sol.maxUniqueSplit("aa"))

File: day71-80/day79/1593. Split a String Into the Max Number of Unique Substrings.php
<?php

class Solution {
    private $maxLength = 1;
    private function backtrack($i, $s, &$seen) {
        if ($i == strlen($s)) {
            $this->maxLength = max($this->maxLength, count($seen));
            return;
        }
        for ($j = $i + 1; $j <= strlen($s); $j++) {
            $substring = substr($s, $i, $j - $i);
            if (in_array($substring, $seen)) {
                continue;
            }
            $seen[] = $substring;
            $this->backtrack($j, $s, $seen);
            array_pop($seen);
        }
    }

    public function maxUniqueSplit($s) {
        $seen = array();  
        $this->backtrack(0, $s, $seen);  
        return $this->maxLength;  
    }
}


$solution = new Solution();
echo $solution->maxUniqueSplit("ababccc");  
